TELCODOCS-640: Optimizing CPU usage (#51562)
openshift/openshift-docs: modules/optimizing-by-encapsulation.adoc

// Module included in the following assemblies:
//
// * scalability_and_performance/optimizing-networking.adoc

[id="optimizing-cpu-usage_{context}"]
= Encapsulating mount namespaces

> Mount namespaces are used to isolate mount points so that processes in different namespaces cannot view each others' files. Encapsulation is the process of moving mount namespaces to an alternative location, known to the owner but hidden from intrusive visitors.

Author: I agree that security is just a potential by-product, so we should not allude to it in the definition of encapsulation right at the start.

> Mount namespaces are used to isolate mount points so that processes in different namespaces cannot view each others' files. Encapsulation is the process of moving mount namespaces to an alternative location, known to the owner but hidden from intrusive visitors.

Author: Changed to "Encapsulation is the process of moving Kubernetes mount namespaces to an alternative location where they will not be constantly scanned by the host operating system."

The host operating system uses `systemd` to constantly scan all mount namespaces: both the standard Linux mounts and the numerous mounts that Kubernetes uses to operate. The current implementation of Kubelet and CRI-O both use the top-level namespace for all container and Kubelet mountpoints. However, encapsulating these container-specific mountpoints in a private namespace reduces `systemd` overhead with no difference in functionality. Using a separate mount namespace for both CRI-O and Kubelet can encapsulate container-specific mounts from any `systemd` or other host OS interaction.

This ability to potentially achieve major CPU optimization is now available to all {product-title} administrators. Encapsulation can also improve security, by storing Kubernetes-specific mount points in a location safe from inspection by unprivileged users.

== How encapsulation works

> The following diagrams illustrate a Kubernetes installation before and after encapsulation. Both scenarios show example containers which have mount propagation settings of bidirectional, host-to-container, and none.

Reviewer: Would be useful to expand on the concept of mount propagation
Which containers have which type etc

Reviewer 2: Maybe just a link to the kubernetes documentation would be better than rehashing the whole thing here? https://kubernetes.io/docs/concepts/storage/volumes/#mount-propagation

Author: Will add a reference to the topic of mount propagation.

Reviewer: I feel there still is a disconnect between what is written in this section and the diagrams
I.e. I  can not easily follow how the text translates into the diagrams in the preview

Reviewer 2: I think the diagram updates will make it clearer; they were just finished yesterday and haven't been updated yet.

Author: With Jim's help, there's been a rewrite of the narrative. When the diagram is complete, we can do another pass.

Author: The finished diagrams are in place now.

> image::before-k8s-mount-propagation.png[Before encapsulation]

Reviewer: The image renders rather blurry in the preview

Author: The images for now are thumbnails from a screenshot. The diagram is still under development by the graphics artist.

> Here we see `systemd`, host OS processes, `kubelet`, and the container runtime pooling access to all mount namespaces. There is no optimisation of CPU usage.

Reviewer: The sentence "There is no optimisation of CPU usage." feel out-of-place.
Suggest moving the entire section to before the image and this the last sentence

Author: Replaced.

Author: Done.

> Here we see `systemd`, host OS processes, `kubelet`, and the container runtime pooling access to all mount namespaces. There is no optimisation of CPU usage.

Reviewer 2: ```suggestion
Here we see `systemd`, host OS processes, `kubelet`, and the container runtime sharing a single mount namespace.

* `systemd`, host OS processes, `kubelet`, and the container runtime each have access to and visibility of all mount points.

* A mount originating in Container 1, such as `/run/a` is visible to systemd, host OS processes, Kubelet, CRI-O, and other containers with host-to-container or bidirectional mount propagation configured.
```

Author: Check edit with Sharat.

Reviewer 2: @sakhoury ^^

Reviewer 3: The suggestion looks good to me @lack @tmulquee!

Author: Verified.

The following diagram illustrates the system state after encapsulation.

> image::after-k8s-mount-propagation.png[After encapsulation]

Reviewer: Blurry in preview

Author: As above.

> * The main `systemd` scanning process is no longer devoted to unnecessary scanning of Kubernetes-specific mount points. It scans systemd-specific and host mountpoints, and is visible to `systemd` instances in the container space.

Reviewer 2: ```suggestion
* The main `systemd` process is no longer devoted to unnecessary scanning of Kubernetes-specific mount points. It only monitors systemd-specific and host mountpoints.
```

I wasn't sure what that last sentence was talking about - what is visible, and what does "systemd instances in the container space" mean?

Author: Done.

Author: Done.

> * The host OS processes access the top-level namespace at `run/host`.

Reviewer 2: ```suggestion
* The host OS processes can access only the systemd and host mountpoints.
```

Author: Done.

Author: Done.

* Using a separate mount namespace for both CRI-O and Kubelet completely segregates all container-specific mounts away from any `systemd` or other host OS interaction whatsoever.

> * In Container A, mounts originating in CRI-O, and Kubelet, and container volume mounts with bidirectional mount propagation are no longer visible to `systemd` or the host OS.

Reviewer 2: ```suggestion
* Container 1, configured with bidirectional mount propagation, can access systemd and host mounts, Kubelet and CRI-O mounts.  A mount originating in Container A, such as `/run/a` is visible to Kubelet, CRI-O, and other containers with host-to-container or bidirectional mount propagation configured (like Container B), but not visible to systemd or host OS processes.
```

Author: Done.

Author: Done.

> * In Container B, mounts originating in the host OS are still visible to CRI-O, Kubelet, and container volume mounts with host-to-container or bidirectional mount propagation.

Reviewer 2: ```suggestion
* Container 2, configured with host-to-container mount propagation, can access systemd and host mounts, Kubelet and CRI-O mounts.  A mount originating in Container 2, such as `/run/b`, is not visible to any other context.
```

Author: Done.

Author: Done.

> * In Container C, mounts originating in CRI-O, Kubelet, and container volume mounts with no propagation are not visible.

Reviewer 2: ```suggestion
* Container 3, configured with no mount propagation, has no visibility of external mount points. A mount originating in Container3, such as `/run/c`, is not visible to any other context.
```

Author: Done.

The host operating system uses `systemd` to constantly scan all mount namespaces: both the standard Linux mounts and the numerous mounts that Kubernetes uses to operate. The current implementation of Kubelet and CRI-O both use the top-level namespace for all container and Kubelet mountpoints. However, encapsulating these container-specific mountpoints in a private namespace reduces `systemd` overhead with no difference in functionality. Using a separate mount namespace for both CRI-O and Kubelet can encapsulate container-specific mounts from any `systemd` or other host OS interaction.

This ability to potentially achieve major CPU optimization is now available to all {product-title} administrators. Encapsulation can also improve security, by storing Kubernetes-specific mount points in a location safe from inspection by unprivileged users.

The following diagrams illustrate a Kubernetes installation before and after encapsulation. Both scenarios show example containers which have mount propagation settings of bidirectional, host-to-container, and none.

> image::before-k8s-mount-propagation.png[Before encapsulation]

Reviewer 2: Replace with the new https://github.com/redhataccess/documentation-svg-assets/blob/master/for-web/OpenShift/290_OpenShift/290_OpenShift_encapsulating_mount_namespaces_1122_before.png

Author: Done.

Here we see systemd, host OS processes, Kubelet, and the container runtime sharing a single mount namespace.

* systemd, host OS processes, Kubelet, and the container runtime each have access to and visibility of all mount points.

> * A mount originating in Container 1, such as `/run/a` is visible to systemd, host OS processes, Kubelet, container runtime, and other containers with host-to-container or bidirectional mount propagation configured.

Reviewer 2: In the "after" part, we have a bullet about Container2 and Container3 as well - but they would be identical to the text in the "after" section.

Author: Unclear about required edit?

Reviewer 2: Do we want the "before" and "after" sections to each have a note about each container, or only when the behavior is different?  It seems odd to me that this first section only has a note about Container 1 (which is my fault, I considered the 2 sections separately at first).

One idea might be to move the points from Container2 and Container3 from the "after" section to the "before" section, and then add a short note for each in the "after" section to just say something like "The behavior of Container 2 and Container 3 is unchanged"...

Author: OK will consider what to do.

Author: Done.

The following diagram illustrates the system state after encapsulation.

> image::after-k8s-mount-propagation.png[After encapsulation]

Reviewer 2: Replace with the new https://github.com/redhataccess/documentation-svg-assets/blob/master/for-web/OpenShift/290_OpenShift/290_OpenShift_encapsulating_mount_namespaces_1122_after.png

* The main `systemd` process is no longer devoted to unnecessary scanning of Kubernetes-specific mount points. It only monitors systemd-specific and host mountpoints.

> * The host OS processes can access only the `systemd` and `host` mountpoints.

Reviewer 2: ```suggestion
* The host OS processes can access only the `systemd` and host mountpoints.
```

Author: We checked the style guides on this. We're standardizing on lower case, no ticks, for kubelet, systemd, host.

Author: Done. Going with no backticks on systemd also.

* Using a separate mount namespace for both CRI-O and Kubelet completely segregates all container-specific mounts away from any `systemd` or other host OS interaction whatsoever.

> * Container 1, configured with bidirectional mount propagation, can access `systemd` and `host` mounts, Kubelet and CRI-O mounts. A mount originating in Container 1, such as `/run/a` is visible to Kubelet, CRI-O, and other containers with host-to-container or bidirectional mount propagation configured (like Container 2), but not visible to systemd or host OS processes.

Reviewer 2: ```suggestion
* Container 1, configured with bidirectional mount propagation, can access `systemd` and host mounts, Kubelet and CRI-O mounts. A mount originating in Container 1, such as `/run/a` is visible to Kubelet, CRI-O, and other containers with host-to-container or bidirectional mount propagation configured (like Container 2), but not visible to systemd or host OS processes.
```

Author: Done.

Author: Done.

> * Container 2, configured with host-to-container mount propagation, can access `systemd` and `host` mounts, Kubelet and CRI-O mounts. A mount originating in Container 2, such as `/run/b`, is not visible to any other context.

Reviewer 2: ```suggestion
* Container 2, configured with host-to-container mount propagation, can access `systemd` and host mounts, Kubelet and CRI-O mounts. A mount originating in Container 2, such as `/run/b`, is not visible to any other context.
```

Author: Done.

// Module included in the following assemblies:
//
// * scalability_and_performance/optimizing-cpu-usage

> 

Author: Done.

[id="optimizing-cpu-usage_{context}"]
= Encapsulating mount namespaces

Mount namespaces are used to isolate mount points so that processes in different namespaces cannot view each others' files. Encapsulation is the process of moving Kubernetes mount namespaces to an alternative location where they will not be constantly scanned by the host operating system.

The host operating system uses systemd to constantly scan all mount namespaces: both the standard Linux mounts and the numerous mounts that Kubernetes uses to operate. The current implementation of Kubelet and CRI-O both use the top-level namespace for all container runtime and Kubelet mountpoints. However, encapsulating these container-specific mountpoints in a private namespace reduces systemd overhead with no difference in functionality. Using a separate mount namespace for both CRI-O and Kubelet can encapsulate container-specific mounts from any systemd or other host OS interaction.

> This ability to potentially achieve major CPU optimization is now available to all {product-title} administrators. Encapsulation can also improve security, by storing Kubernetes-specific mount points in a location safe from inspection by unprivileged users.

Author: Done.

The following diagrams illustrate a Kubernetes installation before and after encapsulation. Both scenarios show example containers which have mount propagation settings of bidirectional, host-to-container, and none.

image::before-k8s-mount-propagation.png[Before encapsulation]

Here we see systemd, host OS processes, Kubelet, and the container runtime sharing a single mount namespace.

* systemd, host OS processes, Kubelet, and the container runtime each have access to and visibility of all mount points.

> * A mount originating in Container 1, such as `/run/a` is visible to systemd, host OS processes, Kubelet, container runtime, and other containers with host-to-container or bidirectional mount propagation configured.

Reviewer 2: @tmulquee I think we need this in addition to the other edit in the "after" section...

* Container 2, configured with host-to-container mount propagation, can access systemd and host mounts, Kubelet and CRI-O mounts. A mount originating in Container 2, such as `/run/b`, is not visible to any other context.

* Container 3, configured with no mount propagation, has no visibility of external mount points. A mount originating in Container 3, such as `/run/c`, is not visible to any other context.

The following diagram illustrates the system state after encapsulation.

image::after-k8s-mount-propagation.png[After encapsulation]

* The main systemd process is no longer devoted to unnecessary scanning of Kubernetes-specific mount points. It only monitors systemd-specific and host mountpoints.

* The host OS processes can access only the systemd and host mountpoints.

> * Using a separate mount namespace for both CRI-O and Kubelet completely segregates all container-specific mounts away from any systemd or other host OS interaction whatsoever.

Author: Done.

> * Container 1, configured with bidirectional mount propagation, can access systemd and host mounts, Kubelet and CRI-O mounts. A mount originating in Container 1, such as `/run/a` is visible to Kubelet, CRI-O, and other containers with host-to-container or bidirectional mount propagation configured (like Container 2), but not visible to systemd or host OS processes.

Author: Done.
